cmd/collectors/zapi/plugins/sensor/sensor.go

```diff
@@ -0,0 +1,295 @@
+							sensorEnvironmentMetricMap[iKey].currentSensor = make(map[string]*sensorValue, 0)
+						}
+						sensorEnvironmentMetricMap[iKey].currentSensor[iKey2] = &sensorValue{name: iKey2, value: value, unit: sensorUnit}
+					}
+				}
+
+			}
+		}
+	}
+
+	for _, k := range eMetrics {
+		my.createEnvironmentMetric(k)
+	}
+	for key, v := range sensorEnvironmentMetricMap {
+		instance, err := my.data.NewInstance(key)
+		instance.SetLabel("node", key)
+		for _, k := range eMetrics {
+			m := my.data.GetMetric(k)
+			switch k {
+			case "power":
+				var sumPower float64
+				if len(v.powerSensor) > 0 {
+					for _, v1 := range v.powerSensor {
+						if v1.unit == "mW" {
+							sumPower += sumPower + (v1.value / 1000)
+						} else if v1.unit == "W" {
+							sumPower += sumPower + v1.value
+						} else {
+							my.Logger.Warn().Str("unit", v1.unit).Str("value", fmt.Sprintf("%f", v1.value)).Msg("unknown power unit")
+						}
+					}
+				} else if len(v.voltageSensor) > 0 && len(v.voltageSensor) == len(v.currentSensor) {
+					// sort voltage keys
+					voltageKeys := make([]string, 0, len(v.voltageSensor))
+					for k := range v.voltageSensor {
+						voltageKeys = append(voltageKeys, k)
+					}
+					sort.Strings(voltageKeys)
+
+					// sort current keys
+					currentKeys := make([]string, 0, len(v.currentSensor))
+					for k := range v.currentSensor {
+						currentKeys = append(currentKeys, k)
+					}
+					sort.Strings(currentKeys)
+
+					for i, _ := range currentKeys {
+						currentKey := currentKeys[i]
+						voltageKey := voltageKeys[i]
+
+						//get values
+						currentSensorValue := v.currentSensor[currentKey]
+						voltageSensorValue := v.voltageSensor[voltageKey]
+
+						// convert units
+						if currentSensorValue.unit == "mA" {
+							currentSensorValue.value = currentSensorValue.value / 1000
+						} else if currentSensorValue.unit == "A" {
+							// do nothing
+						} else {
+							my.Logger.Warn().Str("unit", currentSensorValue.unit).Str("value", fmt.Sprintf("%f", currentSensorValue.value)).Msg("unknown current unit")
+						}
+
+						if voltageSensorValue.unit == "mV" {
+							voltageSensorValue.value = voltageSensorValue.value / 1000
+						} else if voltageSensorValue.unit == "A" {
+							// do nothing
+						} else {
+							my.Logger.Warn().Str("unit", voltageSensorValue.unit).Str("value", fmt.Sprintf("%f", voltageSensorValue.value)).Msg("unknown voltage unit")
+						}
+
+						p := currentSensorValue.value * voltageSensorValue.value
+
+						if strings.Contains(voltageSensorValue.name, "in") || strings.Contains(voltageSensorValue.name, "IN") {
+							if strings.Contains(currentSensorValue.name, "in") || strings.Contains(currentSensorValue.name, "IN") {
+								p = p / 0.93
+							}
+						}
+						sumPower += p
+					}
+				}
+				// convert to KW
+				sumPower = sumPower / 1000
+				err = m.SetValueFloat64(instance, sumPower)
+				if err != nil {
+					my.Logger.Error().Stack().Err(err).Msg("error")
```

Comment:
no need for stack if we use a good message, maybe something like `.Float(sumPower).Msg("Unable to set power")`